R/read_vpts_from_url.R

```diff
@@ -0,0 +1,84 @@
+#' Reading vpts files from a vector of urls
+#'
+#' This is a simple helper to read a vector of urls as vpts files using vroom
+#' and httr2. This wrapper makes use of purrr to create requests that can be
+#' handled in parallel with our own custom retry settings and user agent.
+#'
+#' Apart from parallelisation and these custom settings, this could also be
+#' handled by simple call to `vroom::vroom(file = urls)`
+#'
+#' This function also includes column specifications for the vpts csv data
+#' standard. However, [bioRad::as.vpts()] currently doesn't support factors,
+#' thus any fields sent to that function need to be parsed as character vectors.
+#'
+#' @param urls description
+#'
+#' @return A list of tibbles, one for each url.
+#' @noRd
+#'
+#' @examples
+#' c("https://aloftdata.s3-eu-west-1.amazonaws.com/baltrad/daily/bejab/2024/bejab_vpts_20240305.csv",
+#' "https://aloftdata.s3-eu-west-1.amazonaws.com/baltrad/daily/bejab/2024/bejab_vpts_20240306.csv",
+#' "https://aloftdata.s3-eu-west-1.amazonaws.com/baltrad/daily/bejab/2024/bejab_vpts_20240307.csv"
+#' ) |>
+#'  read_vpts_from_url()
+
+read_vpts_from_url <- function(urls) {
+
+  ## this could also be done by passing the vector of urls to readr::read_csv()
+  ## or vroom::vroom(), but both would be slower because they are not parallel
+  ## and wouldn't declare our custom user agent or allow us to set retry
+  ## criteria
+
+  purrr::map(urls, httr2::request) |>
+    # Identify ourselves in the request
+    purrr::map(req_user_agent_getrad) |>
+    # Set retry conditions
+    purrr::map(~ httr2::req_retry(
+      .x,
+      max_tries = 15, backoff = \(x) sqrt(x) * 2,
+      is_transient = \(resp) resp_status(resp) %in% c(429),
+      retry_on_failure = TRUE
+    )) |>
+    # Perform the requests in parallel
+    httr2::req_perform_parallel() |>
```

Comment:
By performing the requests in parallel, no retries are used in older versions. This seems to have been fixed/changed in a more recent version of httr2.




![image](https://github.com/user-attachments/assets/ceb2f8cc-67cf-460c-9db1-5825a5e9c2d0)

![image](https://github.com/user-attachments/assets/5184257b-4cfc-4adb-9281-b8b342f5f7aa)


https://httr2.r-lib.org/reference/req_perform_parallel.html

Used to be:

![image](https://github.com/user-attachments/assets/1e12bd81-5a08-42dd-8ef3-57c328e88ef8)


Will still work on older versions, but with the old limitations. Consider requiring httr2 1.1.1